### ISSUE TITLE
[GLSL ES 1.0.17] extension directives must occur before any non-preprocessor tokens

### ISSUE THREAD

**Issue by JoeDavisIMG** [KhronosGroup/glslang#97]
In page 15 of the GLSL ES 1.0.17 specification, it states "extension directives must occur before any non-preprocessor tokens".

The following shader should cause a compiler error:
------------
#version 100
uniform mediump vec4 colour;
#extension GL_OES_standard_derivatives : enable
void main() {
    gl_FragColor = colour;
}
------------
In Glslang Version: 3.0.790 15-Oct-2015, an error isn't generated.

**Comment by xorgy** [comment]
I suspect this is common to all GLSL and ESSL versions. It is at least the same in 4.50.

**Comment by xorgy** [comment]
I have a fix for this now, but I also have to fix all the tests which violate this rule (ouch) and make sure they don't behave differently after I put the `#extension` directives at the top of the file.

**Comment by johnkslang** [comment]
Correct.  The gslang "Test" directory heavily depends on this feature, as does much shipping content.  This difference between the spec and what's been happening in practice is an internal issue in Khronos.

There is as reasonable interpretation (depending on that part of the which spec is read) that the #extension granularity could be set by extension specs, and there was a default, but I don't want to rehash the whole discussion here.

In practice, I'm sure the best that can happen is getting a warning from glslang, and existing content does not have to change.

**Comment by xorgy** [comment]
@johnkslang So it should warn but not error, and I should just update the test outputs to reflect that the warning was emitted?

**Comment by johnkslang** [comment]
Yes, I think that is okay, if the change is just warn on detection of #extension after non-PP stuff.
